Add accname tests for “Hidden not referenced” step

adampage opened the pull request:
Closes https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/32

adampage committed:
add tests

adampage commented:
Just getting my feet wet here. Could I trouble you for some initial thoughts, @cookiecrook?

[These failures](https://wpt.fyi/results/accname/name/comp_hidden_not_referenced.html?diff&filter=ADC&run_id=5151534562672640&run_id=5089254785417216) related to `content-visibility: hidden` and `aria-hidden="true"` make me wonder if I’m misunderstanding the accname spec, or if additional language is needed to handle elements that meet [the spec’s definition of “hidden”](https://w3c.github.io/accname/#comp_hidden_not_referenced:~:text=Nodes%20with%20the%20aria%2Dhidden%3D%22true%22%20property%3A%20it%20is%20considered%20hidden%2C%20matching%20the%20%22explicitly%20hidden%22%20guideline.) but that _also remain **focusable**_?  (related: https://github.com/w3c/aria/pull/1856)

cookiecrook commented:
> [error]We stopped hearing from agent Hosted Agent. Verify the agent machine is running and has a healthy network connection. Anything that terminates an agent process, starves it for CPU, or blocks its network access can cause this error. For more information, see: https://go.microsoft.com/fwlink/?linkid=846610
> Agent: Hosted Agent
> Started: Today at 10:31 PM

Well, at least that explains why there are no results for WebKit.

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -11,6 +11,31 @@
+<button class="ex" data-testname="content-visibility:hidden button, labelledby content-visibility:hidden button" data-expectedlabel="" id="Kiwi" style="content-visibility:hidden;" aria-labelledby="Eggplant">Kiwi</button>
+<button class="ex" data-testname="aria-hidden=true button, labelledby aria-hidden=true button" data-expectedlabel="" id="Lychee" aria-hidden="true" aria-labelledby="Fig">Lychee</button>
+
+<!-- Non-hidden buttons which are labelled by hidden buttons -->
```
Non-hidden elements labeled by hidden elements is plausible, but not so much with buttons referencing other hidden buttons... Here are some potential examples:

button "Do it!" with hidden referenced label and another hidden but non-referenced node
```html
<button aria-labelledby="button_label">
  <!-- no rendered text, but maybe a background image or some other anti-pattern -->
  <span hidden id="button_label">Do it!</span>
  <span hidden id="not_referenced"> but this should not be in the label </span>
</button>
```

link "More Info about Widgets" with other referenced and non-referenced hidden content
```html
<a href="#" aria-labelledby="link_text link_hidden_text">
  <span id="link_text">
    More Info
    <span hidden> hidden </span>
  </span>
  <span hidden> hidden </span>
  <span hidden id="link_hidden_text"> about Widgets </span>
</button>
```

@aleventhal or @jaragunde probably wrote a bunch more "hidden not referenced" test cases for Chromium and may be able to share a direct link. 

I also recall discussions about inverting `aria-hidden=true` with a descendant `aria-hidden=false` in a visibly rendered node, and similar for inverting CSS `visibility: hidden;` with `visibility: visible;` in a descendant node, though I can't recall where WebKit, Gecko, and Chromium stand on those edge cases.

cookiecrook replied:
h1 label "not hidden not hidden" IIRC
```html
<h1>not hidden
  <div style="visibility: hidden;">
    hidden not referenced
    <span style="visibility: visible;">not hidden</div>
  </div>
</h1> 
```

ditto for aria-hidden, IIRC
```html
<h1>not hidden
  <div aria-hidden="true">
    hidden not referenced
    <span aria-hidden="false">not hidden</div>
  </div>
</h1> 
```

whatever the result: we've seen web devs try things like this, so would be good to get interop.

cookiecrook replied:
Ooh, and this one. expected label: "not hidden"

```html
<h1>not hidden
  <div hidden>
    hidden not referenced (not rendered)
    <span aria-hidden="false">
        hidden not referenced (aria-hidden=false but still hidden b/c not rendered)
        <!-- there will be some interop differences here -->
    </div>

  </div>
</h1> 
```

cookiecrook commented:
I thought of another example: the collapsed (hidden) contents of summay/details, when a role is applied that gets namefrom contents.

<summary role="comment" data-expectedlabel="visible contents" data-testname="namefrom:contents(comment role) on summary w/ hidden details" class="ex">
  visible contents
  <details>hidden contents</details>
</summary>

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -11,6 +11,31 @@
 </head>
 <body>
 
+<!-- Buttons hidden by assorted techniques -->
```
WebKit's behavior seems more correct to me, sending an error code, but unfortunately the WPT runtime has a bug and [crashes on that error code](https://github.com/web-platform-tests/wpt/issues/39617), resulting in entire set of test results being dropped for that run.

adampage committed:
replace all tests

adampage commented:
> I thought of another example: the collapsed (hidden) contents of summay/details, when a role is applied that gets namefrom contents.
> 
> ```
> <summary role="comment" data-expectedlabel="visible contents" data-testname="namefrom:contents(comment role) on summary w/ hidden details" class="ex">
>   visible contents
>   <details>hidden contents</details>
> </summary>
> ```

My brain broke on this one. 😅  First of all, `<details>` should enclose `<summary>` here, right?  Not the other way around?  And then the intention is to test the accessible name of the nested `summary` when it’s given an explicit role allowing name form contents?

adampage commented:
Hey @cookiecrook, thanks again for your patience, for your feedback, and for all the test case suggestions. With the exception of the `<summary role="comment">` one, I believe I’m all caught up, so I went ahead and resolved your previous comments.

In the case of `aria-hidden="false"` elements nested within `aria-hidden="true"`, it seems neither Chrome nor Safari succesfully _un_-hide the nested content, which appears to put them in conformance with both [ARIA 1.2](https://www.w3.org/TR/wai-aria-1.2/#:~:text=In%20other%20words%2C%20aria%2Dhidden%3D%22true%22%20on%20a%20parent%20overrides%20aria%2Dhidden%3D%22false%22%20on%20descendants.) and the [current editor’s draft](https://w3c.github.io/aria/#aria-hidden:~:text=In%20other%20words%2C%20aria%2Dhidden%3D%22true%22%20on%20a%20parent%20overrides%20aria%2Dhidden%3D%22false%22%20on%20descendants.), but I also see there’s still ongoing conversation in #1256. In any case, I left these WPT tests aspirationally expecting that `aria-hidden="true"` will in fact re-expose that content to the accessibility tree.

One weird behavior I noticed is that one of the two failing cases — [#8](https://github.com/web-platform-tests/wpt/pull/40074/files#diff-af4b22766e9e3d4812289d6ed679a7ee550137e468aae2d80871d1649f528984R104-R114) — actually does appear to _kind of_ work in reality, at least when testing in macOS with VoiceOver (for both Chrome and Safari). When I move the virtual cursor to the rendered heading, I **do** get the expected announcement of “_visible to all users, un-hidden for all users_”. In Chrome, the devtools accessiblity tree shows both intended `StaticText` nodes as exposed, but then it also shows the parent heading as having an accessible name matching just the first node:

<img width="2055" alt="image" src="https://github.com/web-platform-tests/wpt/assets/1395630/bc96eee9-b346-4d01-b274-c0cdd8034f0f">

So that makes me suspicious that perhaps there’s a bug/feature in VoiceOver where it announces the full contents of the heading on focus of the virtual cursor, rather than the computed accessible name? As more interesting evidence, the rotor identifies the heading by the computed accessible name:

![image](https://github.com/web-platform-tests/wpt/assets/1395630/ee561343-d9d3-4e47-8bf3-5eb61023a7a2)

I’ll plan to do a little more testing of that in Windows to see how JAWS and NVDA behave.

adampage committed:
Merge remote-tracking branch 'upstream/master' into accname-comp-hidden-not-referenced

cookiecrook commented:
> First of all, `<details>` should enclose `<summary>` here, right? Not the other way around? And then the intention is to test the accessible name of the nested summary when it’s given an explicit role allowing name form contents?

Yes, sorry. I think this markup is right now, but I'm no longer sure of the expectation.

```html
<details role="comment" data-expectedlabel="visible contents" data-testname="namefrom:contents(comment role) on summary w/ hidden details" class="ex">
  <summary> visible contents </summary>
  hidden contents
</details>
```

Hmm… I'm no longer sure the above is expected, because I think it'd be reasonable for implementations to not provide a backing accessibility node for `<details>` itself when only the `<summary>` is displayed. Probably just add a todo comment.

cookiecrook commented:
> I left these WPT tests aspirationally expecting that aria-hidden="true" will in fact re-expose that content to the accessibility tree.

Since it's still ambiguous, we should probably comment those tests, with a link to the blocking spec issue.

cookiecrook commented:
> that makes me suspicious that perhaps there’s a bug/feature in VoiceOver

Do you have a permanent link to that test case? I can write up a bug and start to determine if it's an engine problem or a VO problem.

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -13,16 +14,119 @@
+<button
+  class="ex"
+  data-expectedlabel="visible to all users"
+  data-testname="#1: button containing unreferenced text nodes that are hidden, visible to all users, and hidden from all users"
```
Let's remove `#1: `, `#2: ` etc from the test names, and we might decide to insert some in the middle, and then our choice is to break the history, or have a misordered list, or insert something silly like `#1.5: `

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -13,16 +14,119 @@
+
+<h2
+  class="ex"
+  data-testname="#7: heading with name from content, containing content that is hidden with nested content that is un-hidden"
```
If any of these are from the examples in the spec, or from examples in the APG, it'd be good to include a link, so future test editors can reference the explanatory comments. 

If these are all new, it'd be good to provide some other explanation. 

In this case, there is a line in the spec (or in an issue) about aria-hidden=false. Link to that if you can fined it.

Nit testname updates because "hidden" was ambiguous, and this behavior differs if the "unhidden" content id not rendered.
```suggestion
  data-testname="heading with name from content, containing rendered content with aria-hidden=true with nested, rendered content with aria-hidden=false"
```

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -13,16 +14,119 @@
+
+<h2
+  class="ex"
+  data-testname="#8: heading with name from content, containing content that is hidden from all users with nested content that is un-hidden for all users"
```
hidden/un-hidden are ambiguous here too. I would recommend referencing `visibility:hidden` and `visibility:visible` directly in the testname for clarity...

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -13,16 +14,119 @@
+  data-expectedlabel="visible to all users"
+  data-testname="#1: button containing unreferenced text nodes that are hidden, visible to all users, and hidden from all users"
+>
+  <span aria-hidden="true">hidden,</span>
```
This is a comment about the label contents… not the testname. I think the testname could explicitly reference the technique used to do the hiding, but the label contents only need reference if they are hidden and from whom.

cookiecrook reviewed (commented):
Tests look good overall... Thanks for all your hard work on this one! 

Minor nits below about testname and label names for clarity and permalink longevity.

adampage committed:
Merge remote-tracking branch 'upstream/master' into accname-comp-hidden-not-referenced

adampage committed:
remove numbers from test names

adampage committed:
disambiguate “hidden” terminology in test names

adampage committed:
add todo comment for future details test

adampage committed:
comment out aria-hidden=false test cases

adampage committed:
finesse test name

adampage committed:
lint fix

adampage commented:
> Do you have a permanent link to that test case? I can write up a bug and start to determine if it's an engine problem or a VO problem.

@cookiecrook well, my premise with this test case was that an element with `visibility:visible` should be expected to override an ancestor’s `visibility:hidden` and surgically re-expose its own contents to the accessibility tree. But now I’m failing to find any spec’d documentation in either ARIA or CSS to support that expectation.

Our 1.3 editors draft [doesn’t seem to make any exceptions](https://w3c.github.io/aria/#:~:text=Elements%2C%20including%20their%20descendent%20elements%2C%20that%20have%20host%20language%20semantics%20specifying%20that%20the%20element%20is%20not%20displayed%2C%20such%20as%20CSS%20display%3Anone%2C%20visibility%3Ahidden%2C%20or%20the%20HTML%20hidden%20attribute.) for descendants of an element with `visibility:hidden`:

> Elements, including their descendent elements, that have host language semantics specifying that the element is not displayed, such as CSS display:none, visibility:hidden, or the HTML hidden attribute.

But as far as interop goes, it does seem like there’s a test case here worth pursuing, since [Firefox _passes_ the test case (as written)](https://wpt.fyi/results/accname/name/comp_hidden_not_referenced.html?diff&filter=ADC&run_id=6450609059790848&run_id=5194013566566400#:~:text=heading%20with%20name%20from%20content%2C%20containing%20element%20that%20is%20visibility%3Ahidden%20with%20nested%20content%20that%20is%20visibility%3Avisible) while [Chrome fails](https://wpt.fyi/results/accname/name/comp_hidden_not_referenced.html?diff&filter=ADC&run_id=5190970582237184&run_id=5110425147932672#:~:text=%2B1-,heading%20with%20name%20from%20content%2C%20containing%20element%20that%20is%20visibility%3Ahidden%20with%20nested%20content%20that%20is%20visibility%3Avisible,-MISSING) and [Safari fails](https://wpt.fyi/results/accname/name/comp_hidden_not_referenced.html?diff&filter=ADC&run_id=5155656690040832&run_id=5100171718819840#:~:text=heading%20with%20name%20from%20content%2C%20containing%20element%20that%20is%20visibility%3Ahidden%20with%20nested%20content%20that%20is%20visibility%3Avisible).

A strict reading of that language in the ARIA 1.3 spec suggests that I should keep the test case but change the expected label to “_visible to all users,_”, then file a Mozilla bug?

adampage commented:
@cookiecrook besides [this last comment](https://github.com/web-platform-tests/wpt/pull/40074#issuecomment-1742323982), I believe I’ve addressed all your feedback with my latest batch of commits. Please let me know what you think?

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -13,16 +14,145 @@
+  <span>visible to all users</span>
+</button>
+
+<!-- TODO: Restore the following test once expectation for `aria-hidden="false"`
+<!--       behavior is documented. See: -->
+<!--       https://github.com/w3c/aria/issues/1256 -->
```
I know you worked hard on these, but I think it would be better to check in the patch without the commented code, and only add specific examples once the new issue came to resolution... Otherwise someone might see 1256 is resolved (in the future) and uncomment these even if they don't align with the resolution… Or also bad, these could just sit around commented indefinitely.

cookiecrook replied:
maybe just keep your test names, since they should be clear enough to show what needs to be tested? such as:

```html
<!-- 
  Todo: once https://github.com/w3c/aria/issues/1256 resolved.

  - button labelled by an element that is aria-hidden=true which contains a nested child that is aria-hidden=false 
  - test name 2
  - test name 3
  - etc.
-->
```

cookiecrook replied:
thanks for working so hard on these!

cookiecrook replied:
Acknowledging that **August James** asked you to comment them, and **October James** told you to remove them, so I won't stop you from squash merging as-is... Take your pick. Thanks.

adampage replied:
Ha ha, no trouble at all, @cookiecrook  — happy for the opportunity to contribute at all. I went ahead and removed the test cases that are blocked by [aria#1256](https://github.com/w3c/aria/issues/1256) but left a comment identifying each one by name. 👍🏻 

Based on our quick conversation in the WPT meeting, I also left the expectation as is for the `...visibility:hidden with nested content that is visibility:visible` test case, so Firefox is currently the only pass for that one.

cookiecrook reviewed (approved):
Approving either way, but with a minor preference for removing the commented code.

adampage committed:
Merge remote-tracking branch 'upstream/master' into accname-comp-hidden-not-referenced

adampage committed:
remove test cases blocked by aria #1256
